packages/rfw/README.md

```diff
@@ -208,7 +328,409 @@ class _ExampleState extends State<Example> {
 }
 ```
 
-## Fetching remote widget libraries remotely
+Widgets in local widget libraries are represented by closures that are
+invoked by the runtime whenever a local widget is referenced.
+
+The closure uses the
+[LocalWidgetBuilder](https://pub.dev/documentation/rfw/latest/rfw/LocalWidgetBuilder.html)
+signature. Like any builder in Flutter, it takes a 
+[`BuildContext`](https://api.flutter.dev/flutter/widgets/BuildContext-class.html),
+which can be used to look up inherited widgets. For example, when
+using the `createCoreWidgets` widgets,
+[`AnimationDefaults.curveOf`](https://pub.dev/documentation/rfw/latest/rfw/AnimationDefaults/curveOf.html)
+and
+[`AnimationDefaults.durationOf`](https://pub.dev/documentation/rfw/latest/rfw/AnimationDefaults/durationOf.html)
+can be used with the given `BuildContext` to get default curve and
+duration settings for animations.
+
+The other argument is a [`DataSource`](https://pub.dev/documentation/rfw/latest/rfw/DataSource-class.html).
+This gives access to the arguments that were provided to the widget in
+the remote widget library.
+
+For example, consider the example above, where the remote widget library is:
+
+```
+import local;
+widget root = GreenBox(
+  child: Hello(name: "World"),
+);
+```
+
+The `GreenBox` widget is invoked with one argument (`child`), and the
+`Hello` widget is invoked with one argument (`name`).
+
+In the definitions of `GreenBox` and `Hello`, the data source is used
+to pull out these arguments.
+
+### Obtaining arguments from the `DataSource`
+
+The arguments are a tree of maps and lists with leaves that are Dart
+scalar values (`int`, `double`, `bool`, or `String`), further widgets,
+or event handlers.
+
+#### Scalars
+
+Here is an example of a more elaborate widget argument:
+
+```
+widget root = Foo(
+  bar: { quux: [ 'apple', 'banana', 'cherry' ] },
+);
+```
+
+To obtain a scalar value from the arguments, the
+[DataSource.v](https://pub.dev/documentation/rfw/latest/rfw/DataSource/v.html)
+method is used. This method takes a list of keys (strings or integers)
+that denote the path to scalar in question. For instance, to obtain
+"cherry" from the example above, the keys would be `bar`, `quux`, and
+2, as in:
+
+```dart
+    'Foo': (BuildContext context, DataSource source) {
+      return Text(source.v<String>(['bar', 'quux', 2])!);
+    },
+```
+
+The `v` method is generic, with a type argument that specifies the
+expected type (one of `int`, `double`, `bool`, or `String`). When the
+value of the argument in the remote widget library does not match the
+specified (or inferred) type given to `v`, or if the specified keys
+don't lead to a value at all, it returns null instead.
+
+#### Maps and lists
+
+The `LocalWidgetBuilder` callback can inspect keys to see if they are
+maps or lists before attempting to use them. This is useful to avoid
+potentially expensive work, e.g. iterating through a list that isn't
+even there.
```

Comment:
Changed to "pulling a bunch of items from a list that isn't even there". For example, if you are given an index and a list, and are going to pull many properties from the indexth object in the list, but the list is not there, it'll be more expensive than doing nothing, because for each property you'll try to load the list but it won't be there.

Reply:
Hmm, I see. Even then, though: in that example, I feel like what I'd ideally want to check isn't whether the list is there but whether the indexth element in the list is there. (Which I could do with `isMap`, right? If I'm looking for properties on it via a `DataSource`, I think that means I'm expecting it to be a map.)

Perhaps s/list/map/? So "pulling a bunch of items from a map that isn't even there".

Ah, in fact it looks like that's the direction you take the example in the `DataSource` doc:
> The [isList](https://pub.dev/documentation/rfw/latest/rfw/DataSource/isList.html) and [isMap](https://pub.dev/documentation/rfw/latest/rfw/DataSource/isMap.html) methods can be used to avoid inspecting keys that may not be present (e.g. before reading 15 keys in a map that isn't even present, consider checking if the map is present using [isMap](https://pub.dev/documentation/rfw/latest/rfw/DataSource/isMap.html) and short-circuiting the key lookups if it is not).

Reply:
certainly that's a use case for isMap, and i expect that to be used more than isList.

Reply:
i've changed this prose to talk about isMap instead.